### PR TITLE
Migrate IndexShardOperationPermits to proper executor

### PR DESCRIPTION
**Pull request by DaveCTurner**
Relates #99392, #97879

### CODE REVIEW

**Comment by DaveCTurner** [comment]
This has no impact on any APIs called by serverless so should be ok to merge without a deprecation dance. But I haven't run the tests yet and there may be some mocks in the way.

File: server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java [hunk 1]
```diff
@@ -422,7 +422,7 @@ public void onFailure(Exception e) {
             cancellableThreads.checkForCancel();
             ensureNotRelocatedPrimary(primary);
             action.accept(l2);
-        })), ThreadPool.Names.GENERIC);
+        })), primary.getThreadPool().generic());
```
Comment (DiannaHohensee) [inline server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java:425]: nice :)

**Review by DiannaHohensee — APPROVED** [review]
LGTM! Thank you for putting up a patch.

**Comment by DaveCTurner** [comment]
@elasticmachine please run elasticsearch-ci/bwc
@elasticmachine please run elasticsearch-ci/part-1

**Comment by DaveCTurner** [comment]
Hurrah, merged at last. @DiannaHohensee if you merge `main` into #99392 then you'll get some conflicts but resolving them should reduce the size of the remaining changes.

**Pull request merged**